### ISSUE TITLE
Using Criteria Object

### ISSUE THREAD

**Issue by CrossSlide** [MarimerLLC/cslaforum#447]
I'm trying to use a Criteria object to pass some params (Encapsulated Invocation). I setup my interface class with Fetch(object myCriteria). But when I reach my DataAccess layer class I can't cast the object back to myCriteria to retrieve my values. 

It can't find the namespace of myCriteria. If I try to add the namespace to the DataAccess project I get a circular ref error. What am I doing wrong? Thanks

**Comment by rockfordlhotka** [comment]
I generally don't pass the criteria object into my DAL - I don't want that level of coupling. Instead I have my DAL methods accept discrete values necessary to do the work.

But you could invert that, and have the DAL reference the business library (which means the business library can't reference the DAL) - in which case you can pass the criteria into the DAL.

Either way, I typically have a _separate_ assembly that contains the DAL interface definitions so the business layer can reference that interface assembly, and not directly reference the DAL assembly. Then I use dynamic assembly loading to load the appropriate DAL at runtime.

**Comment by CrossSlide** [comment]
Thank Rocky!

**Comment by ck-g** [comment]
Yes, Fetch takes only zero or one parameter. If you need more than one param you could define a criteria class in your bo. The criteria class is a container for two or more primitive values (see p. 99 in the overview ebook). 

For a sample have a look at p. 98 in the data portal ebook or at the DataPortal_Fetch(UsernameCriteria criteria) method in the PTIdentity class of the ProjectTracker sample: /Samples/ProjectTracker/ProjectTracker.BusinessLibrary.Shared/Security/PTIdentity.cs
- the Csla.Security.UsernameCriteria class has two properties: username, password
- the bo uses this UsernameCriteria in it's Fetch method to call the dal with two discrete values (criteria.username, criteria.password)
- as the dal has no knowledge about the business lib there is no circular reference problem

**Comment by CrossSlide** [comment]
Got it! Thanks @ck-g
